synthtool/gcp/gapic_bazel.py

```diff
@@ -0,0 +1,236 @@
+            raise RuntimeError(
+                f"Unable to generate {service}, the googleapis repository"
+                "is unavailable."
+            )
+
+        # Determine where the protos we are generating actually live.
+        # We can sometimes (but not always) determine this from the service
+        # and version; in other cases, the user must provide it outright.
+        if proto_path:
+            proto_path = Path(proto_path)
+            if proto_path.is_absolute():
+                proto_path = proto_path.relative_to("/")
+        else:
+            proto_path = Path("google/cloud") / service / version
+
+        # Determine bazel target based on per-language patterns
+        # Java:    google-cloud-{{assembly_name}}-{{version}}-java
+        # Go:      gapi-cloud-{{assembly_name}}-{{version}}-go
+        # Python:  {{assembly_name}}-{{version}}-py
+        # PHP:     google-cloud-{{assembly_name}}-{{version}}-php
+        # Node.js: {{assembly_name}}-{{version}}-nodejs
+        # Ruby:    google-cloud-{{assembly_name}}-{{version}}-ruby
+        # C#:      google-cloud-{{assembly_name}}-{{version}}-csharp
+        if bazel_target is None:
+            parts = list(proto_path.parts)
+            while len(parts) > 0 and parts[0] != "google":
+                parts.pop(0)
+            if len(parts) == 0:
+                raise RuntimeError(
+                    f"Cannot determine bazel_target from proto_path {proto_path}."
+                    "Please set bazel_target explicitly."
+                )
+            if language == "python":
+                suffix = f"{service}-{version}-py"
+            elif language == "nodejs":
+                suffix = f"{service}-{version}-nodejs"
+            elif language == "go":
+                suffix = f"gapi-{'-'.join(parts[1:])}-go"
+            else:
+                suffix = f"{'-'.join(parts)}-{language}"
+            bazel_target = f"//{os.path.sep.join(parts)}:{suffix}"
+
+        # Sanity check: Do we have protos where we think we should?
+        if not (googleapis / proto_path).exists():
+            raise FileNotFoundError(
+                f"Unable to find directory for protos: {(googleapis / proto_path)}."
+            )
+        if not tuple((googleapis / proto_path).glob("*.proto")):
+            raise FileNotFoundError(
+                f"Directory {(googleapis / proto_path)} exists, but no protos found."
+            )
+        if not (googleapis / proto_path / "BUILD.bazel"):
+            raise FileNotFoundError(
+                f"File {(googleapis / proto_path / 'BUILD.bazel')} does not exist."
+            )
+
+        # Ensure the desired output directory exists.
+        # If none was provided, create a temporary directory.
+        if not output_dir:
+            output_dir = tempfile.mkdtemp()
+        output_dir = Path(output_dir).resolve()
+
+        # Let's build some stuff now.
+        cwd = os.getcwd()
+        os.chdir(str(googleapis))
+
```

Comment:
Use a try finally to make sure we chdir back into the original directory even if an exception is thrown.